python/cuml/test/utils.py

```diff
@@ -29,12 +29,20 @@
 import pytest
 
 
-def array_equal(a, b, tol=1e-4, with_sign=True):
+def array_equal(a, b, unit_tol=1e-4, total_tol=1e-4, with_sign=True):
+    """
+    Utility function to compare 2 numpy arrays. Two individual elements
+    are assumed equal if they are within `unit_tol` of each other, and two
+    arrays are considered equal if less than `total_tol` percentage of
+    elements are different.
+
+    """
     a = to_nparray(a)
     b = to_nparray(b)
     if not with_sign:
         a, b = np.abs(a), np.abs(b)
-    res = np.max(np.abs(a-b)) < tol
+    print((np.sum(np.abs(a-b) > unit_tol)) / len(a))
+    res = (np.sum(np.abs(a-b) > unit_tol)) / len(a) < total_tol
```

Comment:
should we smoothly handle the case of zero-length arrays since this is such a core utility?
Just add `if len(a) == 0 and len(b) == 0: return True` near the top